src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1056,7 +1056,8 @@ intptr_t ObjectSynchronizer::FastHashCode(Thread* current, oop obj) {
       }
       // Fall thru so we only have one place that installs the hash in
       // the ObjectMonitor.
-    } else if (LockingMode == LM_LEGACY && mark.has_locker() && current->is_lock_owned((address)mark.locker())) {
+    } else if (LockingMode == LM_LEGACY && mark.has_locker()
+               && JavaThread::cast(current)->is_lock_owned((address)mark.locker())) {
```

Comment:
This looks risky.  How about guarding it with a check for current->is_Java_thread()?

Reply:
Yes!